supertokens_python/recipe/session/utils.py

```diff
@@ -391,21 +401,42 @@ def validate_and_normalise_user_input(
     cookie_domain = (
         normalise_session_scope(cookie_domain) if cookie_domain is not None else None
     )
-    top_level_api_domain = app_info.top_level_api_domain
-    top_level_website_domain = app_info.top_level_website_domain
 
-    api_domain_scheme = get_url_scheme(app_info.api_domain.get_as_string_dangerous())
-    website_domain_scheme = get_url_scheme(
-        app_info.website_domain.get_as_string_dangerous()
-    )
-    if cookie_same_site is not None:
-        cookie_same_site = normalise_same_site(cookie_same_site)
-    elif (top_level_api_domain != top_level_website_domain) or (
-        api_domain_scheme != website_domain_scheme
-    ):
-        cookie_same_site = "none"
-    else:
-        cookie_same_site = "lax"
+    async def cookie_same_site_func(
+        req: BaseRequest, user_context: Any
+    ) -> Literal["lax", "none", "strict"]:
+        origin = app_info.website_domain
+        origin_string = origin.get_as_string_dangerous()
+        origin_scheme = get_url_scheme(origin_string)
+        top_level_origin = get_top_level_domain_for_same_site_resolution(origin_string)
+
+        api_domain_scheme = get_url_scheme(
+            app_info.api_domain.get_as_string_dangerous()
+        )
+        top_level_api_domain = get_top_level_domain_for_same_site_resolution(
+            app_info.api_domain.get_as_string_dangerous()
+        )
+
+        cookie_same_site_normalize = "none"
+
+        if cookie_same_site is not None:
+            if isinstance(cookie_same_site, types.FunctionType):
+                cookie_same_site_val = await cookie_same_site(req, user_context)
+                cookie_same_site_normalize = normalise_same_site(cookie_same_site_val)
+            else:
+                cookie_same_site_normalize = normalise_same_site(str(cookie_same_site))
+
+        if (
+            top_level_api_domain != top_level_origin
+            or api_domain_scheme != origin_scheme
+        ):
+            cookie_same_site_ret = "none"
+        else:
+            cookie_same_site_ret = "lax"
+
+        if cookie_same_site is not None:
+            return cookie_same_site_normalize
+        return cookie_same_site_ret
```

Comment:
```suggestion
        return result
```

*_ret* is kinda bad for being part of var name.